File: android/src/main/kotlin/unofficial/twilio/flutter/twilio_unofficial_programmable_video/RoomListener.kt
package unofficial.twilio.flutter.twilio_unofficial_programmable_video

import com.twilio.video.CameraCapturer
import com.twilio.video.ConnectOptions
import com.twilio.video.LocalAudioTrack
import com.twilio.video.LocalAudioTrackPublication
import com.twilio.video.LocalParticipant
import com.twilio.video.LocalVideoTrack
import com.twilio.video.LocalVideoTrackPublication
import com.twilio.video.RemoteParticipant
import com.twilio.video.Room
import com.twilio.video.TwilioException
import com.twilio.video.VideoCapturer

class RoomListener(private var internalId: Int, var connectOptions: ConnectOptions) : BaseListener(), Room.Listener {
    var room: Room? = null

    override fun onConnectFailure(room: Room, e: TwilioException) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onConnectFailure => room sid is '${room.sid}', exception is $e")
        sendEvent("connectFailure", mapOf("room" to roomToMap(room)), e)
    }

    override fun onConnected(room: Room) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onConnected => room sid is '${room.sid}'")
        sendEvent("connected", mapOf("room" to roomToMap(room)))
        room.remoteParticipants.forEach { it.setListener(TwilioUnofficialProgrammableVideoPlugin.remoteParticipantListener) }
    }

    override fun onDisconnected(room: Room, e: TwilioException?) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onDisconnected => room sid is '${room.sid}', exception is $e")
        sendEvent("disconnected", mapOf("room" to roomToMap(room)), e)
    }

    override fun onParticipantConnected(room: Room, remoteParticipant: RemoteParticipant) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onParticipantConnected => room sid is '${room.sid}', remoteParticipant sid is '${remoteParticipant.sid}'")
        sendEvent("participantConnected", mapOf("room" to roomToMap(room), "remoteParticipant" to RemoteParticipantListener.remoteParticipantToMap(remoteParticipant)))
        remoteParticipant.setListener(TwilioUnofficialProgrammableVideoPlugin.remoteParticipantListener)
    }

    override fun onParticipantDisconnected(room: Room, remoteParticipant: RemoteParticipant) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onParticipantDisconnected => room sid is '${room.sid}', participant sid is '${remoteParticipant.sid}'")
        sendEvent("participantDisconnected", mapOf("room" to roomToMap(room), "remoteParticipant" to RemoteParticipantListener.remoteParticipantToMap(remoteParticipant)))
    }

    override fun onReconnected(room: Room) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onReconnected => room sid is '${room.sid}'")
        sendEvent("reconnected", mapOf("room" to roomToMap(room)))
    }

    override fun onReconnecting(room: Room, e: TwilioException) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onReconnecting => room sid is '${room.sid}', exception is $e")
        sendEvent("connected", mapOf("room" to roomToMap(room)), e)
    }

    override fun onRecordingStarted(room: Room) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onRecordingStarted => room sid is '${room.sid}'")
        sendEvent("recordingStarted", mapOf("room" to roomToMap(room)))
    }

    override fun onRecordingStopped(room: Room) {
        TwilioUnofficialProgrammableVideoPlugin.debug("RoomListener.onRecordingStopped => room sid is '${room.sid}'")
        sendEvent("recordingStopped", mapOf("room" to roomToMap(room)))
    }

    private fun remoteParticipantsToList(remoteParticipants: List<RemoteParticipant>): List<Map<String, Any?>> {
        return remoteParticipants.map { RemoteParticipantListener.remoteParticipantToMap(it) }
    }

    private fun roomToMap(room: Room): Map<String, Any?> {
        return mapOf(
                "sid" to room.sid,
                "name" to room.name,
                "state" to room.state.toString(),
                "mediaRegion" to room.mediaRegion,
                "localParticipant" to localParticipantToMap(room.localParticipant),
                "remoteParticipants" to remoteParticipantsToList(room.remoteParticipants)
        )
    }

    private fun localParticipantToMap(localParticipant: LocalParticipant?): Map<String, Any?> {
        val localAudioTrackPublications = localParticipant?.localAudioTracks?.map { localAudioTrackPublicationToMap(it) }
//        val localDataTrackPublications = localParticipant?.localDataTracks?.map { localDataTrackPublicationToMap(it) }
        val localVideoTrackPublications = localParticipant?.localVideoTracks?.map { localVideoTrackPublicationToMap(it) }
        return mapOf(
                "identity" to localParticipant?.identity,
                "sid" to localParticipant?.sid,
                "signalingRegion" to localParticipant?.signalingRegion,
                "networkQualityLevel" to localParticipant?.networkQualityLevel.toString(),
                "localAudioTrackPublications" to localAudioTrackPublications,
                "localVideoTrackPublications" to localVideoTrackPublications
        )
    }

    private fun localAudioTrackPublicationToMap(localVideoTrackPublication: LocalAudioTrackPublication): Map<String, Any> {
        return mapOf(
                "sid" to localVideoTrackPublication.trackSid,
                "localAudioTrack" to localAudioTrackToMap(localVideoTrackPublication.localAudioTrack)
        )
    }

    private fun localAudioTrackToMap(localAudioTrack: LocalAudioTrack): Map<String, Any> {
        return mapOf(
                "name" to localAudioTrack.name,
                "enabled" to localAudioTrack.isEnabled
        )
    }

    private fun localVideoTrackPublicationToMap(localVideoTrackPublication: LocalVideoTrackPublication): Map<String, Any> {
        return mapOf(
                "sid" to localVideoTrackPublication.trackSid,
                "localVideoTrack" to localVideoTrackToMap(localVideoTrackPublication.localVideoTrack)
        )
    }

    private fun localVideoTrackToMap(localVideoTrack: LocalVideoTrack): Map<String, Any> {
        return mapOf(
                "name" to localVideoTrack.name,
                "enabled" to localVideoTrack.isEnabled,
                "videoCapturer" to videoCapturerToMap(localVideoTrack.videoCapturer)
        )
    }

    private fun videoCapturerToMap(videoCapturer: VideoCapturer): Map<String, Any> {
        if (videoCapturer is CameraCapturer) {
            return mapOf(
                    "type" to "CameraCapturer",
                    "cameraSource" to videoCapturer.cameraSource.toString()
            )
        }
        return mapOf(
                "type" to "Unknown",
                "isScreencast" to videoCapturer.isScreencast
        )
    }
}

File: ios/Classes/RoomListener.swift
import Flutter
import Foundation
import TwilioVideo

class RoomListener: BaseListener, RoomDelegate {
    public var connectOptions: ConnectOptions

    public var room: Room?

    init(_ roomId: Int, _ connectOptions: ConnectOptions) {
        self.connectOptions = connectOptions
    }

    func roomDidFailToConnect(room: Room, error: Error) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomDidFailToConnect => room sid is '\(room.sid)', error is \(error)")
        sendEvent("connectFailure", data: [ "room": roomToDict(room) ], error: error)
    }

    func roomDidConnect(room: Room) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomDidConnect => room sid is '\(room.sid)'")
        sendEvent("connected", data: [ "room": roomToDict(room) ])

        for remoteParticipant in room.remoteParticipants {
            remoteParticipant.delegate = SwiftTwilioProgrammableVideoPlugin.remoteParticipantListener
        }
    }

    func roomDidDisconnect(room: Room, error: Error?) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomDidDisconnect => room sid is '\(room.sid)', error is \(String(describing: error))")
        sendEvent("disconnected", data: [ "room": roomToDict(room) ], error: error)
    }

    func roomDidReconnect(room: Room) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomDidReconnect   => room sid is '\(room.sid)'")
        sendEvent("reconnected", data: [ "room": roomToDict(room) ])

        for remoteParticipant in room.remoteParticipants {
            remoteParticipant.delegate = SwiftTwilioProgrammableVideoPlugin.remoteParticipantListener
        }
    }

    func roomIsReconnecting(room: Room, error: Error) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomIsReconnecting => room sid is '\(room.sid)', error is \(error)")
        sendEvent("reconnecting", data: [ "room": roomToDict(room) ], error: error)
    }

    func roomDidStartRecording(room: Room) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomDidStartRecording => room sid is '\(room.sid)'")
        sendEvent("recordingStarted", data: [ "room": roomToDict(room) ])
    }

    func roomDidStopRecording(room: Room) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.roomDidStopRecording => room sid is '\(room.sid)'")
        sendEvent("recordingStopped", data: [ "room": roomToDict(room) ])
    }

    func participantDidConnect(room: Room, participant: RemoteParticipant) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.participantDidConnect => room sid is '\(room.sid)', remoteParticipant sid is '\(String(describing: participant.sid))'")
        sendEvent("participantConnected", data: [ "room": roomToDict(room) as Any, "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant) ])
        participant.delegate = SwiftTwilioProgrammableVideoPlugin.remoteParticipantListener
    }

    func participantDidDisconnect(room: Room, participant: RemoteParticipant) {
        SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.participantDidDisconnect => room sid is '\(room.sid)', remoteParticipant sid is '\(String(describing: participant.sid))'")
        sendEvent("participantDisconnected", data: [ "room": roomToDict(room) as Any, "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant) as Any ])
    }

    func dominantSpeakerDidChange(room: Room, participant: RemoteParticipant?) {
        if let participant = participant {
            SwiftTwilioProgrammableVideoPlugin.debug("RoomListener.dominantSpeakerDidChange => room sid is '\(room.sid)', dominantSpeaker sid is '\(String(describing: participant.sid))'")
            sendEvent("dominantSpeakerChanged", data: [ "room": roomToDict(room) as Any, "remoteParticipant": RemoteParticipantListener.remoteParticipantToDict(participant) as Any ])
        }
    }

    private func remoteParticipantsToArray(_ remoteParticipants: [RemoteParticipant]) -> [[String: Any]] {
        return remoteParticipants.map({ (it) -> [String: Any] in
            return RemoteParticipantListener.remoteParticipantToDict(it)
        })
    }

    private func roomToDict(_ room: Room) -> [String: Any] {
        var roomState: String
        switch room.state {
            case Room.State.connecting:
                roomState = "CONNECTING"
            case Room.State.connected:
                roomState = "CONNECTED"
            case Room.State.reconnecting:
                roomState = "RECONNECTING"
            case Room.State.disconnected:
                roomState = "DISCONNECTED"
            default:
                roomState = "UNKNOWN"
        }

        var dict = [
           "sid": room.sid,
           "name": room.name,
           "state": roomState,
           "mediaRegion": room.mediaRegion as Any,
           "localParticipant": localParticipantToDict(room.localParticipant) as Any,
           "remoteParticipants": remoteParticipantsToArray(room.remoteParticipants)
       ]

       if room.dominantSpeaker != nil {
            dict["dominantSpeaker"] = RemoteParticipantListener.remoteParticipantToDict(room.dominantSpeaker!)
       }

       return dict
    }

    private func localParticipantToDict(_ localParticipant: LocalParticipant?) -> [String: Any]? {
        if let localParticipant = localParticipant {
            var networkQualityLevel: String
            switch localParticipant.networkQualityLevel {
                case NetworkQualityLevel.zero:
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_ZERO"
                case NetworkQualityLevel.one:
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_ONE"
                case NetworkQualityLevel.two:
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_TWO"
                case NetworkQualityLevel.three:
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_THREE"
                case NetworkQualityLevel.four:
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_FOUR"
                case NetworkQualityLevel.five:
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_FIVE"
                default: // or NetworkQualityLevel.unknown
                    networkQualityLevel = "NETWORK_QUALITY_LEVEL_UNKNOWN"
            }

            let localAudioTrackPublications = localParticipant.localAudioTracks.map({ (it) -> [String: Any] in
                return localAudioTrackPublicationToDict(it)
            })

            let localVideoTrackPublications = localParticipant.localVideoTracks.map({ (it) -> [String: Any] in
                return localVideoTrackPublicationToDict(it)
            })

            return [
                "identity": localParticipant.identity,
                "sid": localParticipant.sid as Any,
                "signalingRegion": localParticipant.signalingRegion,
                "networkQualityLevel": networkQualityLevel,
                "localAudioTrackPublications": localAudioTrackPublications,
                "localVideoTrackPublications": localVideoTrackPublications
            ]
        }
        return nil
    }

    private func localAudioTrackPublicationToDict(_ localAudioTrackPublication: LocalAudioTrackPublication) -> [String: Any] {
        return [
            "sid": localAudioTrackPublication.trackSid,
            "localAudioTrack": localAudioTrackToDict(localAudioTrackPublication.localTrack) as Any
        ]
    }

    private func localAudioTrackToDict(_ localAudioTrack: LocalAudioTrack?) -> [String: Any]? {
        if let localAudioTrack = localAudioTrack {
            return [
                "name": localAudioTrack.name,
                "enabled": localAudioTrack.isEnabled
            ]
        }
        return nil
    }

    private func localVideoTrackPublicationToDict(_ localVideoTrackPublication: LocalVideoTrackPublication) -> [String: Any] {
        return [
            "sid": localVideoTrackPublication.trackSid,
            "localVideoTrack": localVideoTrackToDict(localVideoTrackPublication.localTrack) as Any
        ]
    }

    private func localVideoTrackToDict(_ localVideoTrack: LocalVideoTrack?) -> [String: Any]? {
        if let localVideoTrack = localVideoTrack {
            return [
                "name": localVideoTrack.name,
                "enabled": localVideoTrack.isEnabled,
                "videoCapturer": RoomListener.videoSourceToDict(localVideoTrack.source)
            ]
        }
        return nil
    }

    public static func videoSourceToDict(_ videoSource: VideoSource?) -> [String: Any] {
        if let cameraSource = videoSource as? CameraSource {
            var cameraSourceType: String
            switch cameraSource.device?.position {
                case .front:
                    cameraSourceType = "FRONT_CAMERA"
                case .back:
                    cameraSourceType = "BACK_CAMERA"
                default:
                    cameraSourceType = "UNKNOWN"
            }

            return [
                "type": "CameraCapturer",
                "cameraSource": cameraSourceType
            ]
        }
        return [
            "type": "Unknown",
            "isScreenCast": videoSource?.isScreencast ?? false
        ]
    }
}
